keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -3,703 +3,891 @@
+    // +optional
+    BucketClaim corev1.ObjectReference
+
+    // Protocols are the set of data APIs this bucket is expected to support.
+    // The possible values for protocol are:
+    // -  S3: Indicates Amazon S3 protocol
+    // -  Azure: Indicates Microsoft Azure BlobStore protocol
+    // -  GCS: Indicates Google Cloud Storage protocol
+    Protocols []Protocol
+
+    // Parameters is an opaque map for passing in configuration to a driver
+    // for creating the bucket
+    // +optional
+    Parameters map[string]string
+
+    // ExistingBucketID is the unique id of the bucket in the OSP. This field should be
+    // used to specify a bucket that has been created outside of COSI.
+    // This field will be empty when the Bucket is dynamically provisioned by COSI.
+    // +optional
+    ExistingBucketID string
+  }
+
+  Status BucketStatus {
+    // BucketReady is a boolean condition to reflect the successful creation
+    // of a bucket.
+    BucketReady bool
+
+    // BucketID is the unique id of the bucket in the OSP. This field will be
+    // populated by COSI.
+    // +optional
+    BucketID string
+  }
+}
 ```
-message ProvisionerCreateBucketRequest {    
-    // This field is REQUIRED
-    string bucket_name = 1;
 
-    map<string,string> bucket_context = 2;
+## BucketClaim
+
+A claim to create Bucket. BucketClaim is namespace-scoped
+
+```yaml
+BucketClaim {
+  TypeMeta
+  ObjectMeta
 
-    enum AnonymousBucketAccessMode {
-	PRIVATE = 0;
-	PUBLIC_READ_ONLY = 1;
-	PUBLIC_WRITE_ONLY = 2;
-	PUBLIC_READ_WRITE = 3;
-    }
+  Spec BucketClaimSpec {
+    // Name of the BucketClass
+    BucketClassName string
     
-    AnonymousBucketAccessMode anonymous_bucket_access_mode = 3;
-}
+    // Name of a bucket object that was manually 
+    // created to import a bucket created outside of COSI
+    // +optional
+    ExistingBucketName string
+  }
+
+  Status BucketClaimStatus {
+    // BucketReady indicates that the bucket is ready for consumpotion
+    // by workloads
+    BucketReady bool
+
+    // BucketName is the name of the provisioned Bucket in response
+    // to this BucketClaim. It is generated and set by the COSI controller 
+    // before making the creation request to the OSP backend.
+    // +optional
+    BucketName string
+  }
+```
+## BucketClass
 
-message ProvisionerCreateBucketResponse {
-    // Intentionally left blank
+Resouce for configuring common properties for multiple Buckets. BucketClass is cluster-scoped.
+
+```yaml
+BucketClass {
+  TypeMeta
+  ObjectMeta
+
+  // Provisioner is the name of driver associated with this bucket
+  Provisioner string
+
+  // Protocols are the set of data API this bucket is expected to support.
+  // The possible values for protocol are:
+  // -  S3: Indicates Amazon S3 protocol
+  // -  Azure: Indicates Microsoft Azure BlobStore protocol
+  // -  GCS: Indicates Google Cloud Storage protocol
+  Protocols []Protocol
```

Comment:
Is the intention of having protocols here so that an admin can restrict what protocols are exposed? Is that even possible?

Reply:
pending question

Reply:
The intention originally was to use this as a mechanism to specify what protocols a class of buckets would support. On further discussion, we have moved away from this approach and moved the `protocols` field to the BucketClaim - where the application writer gets to specify the protocol they need for their app